File: freight-triage-polyglot-ledger-recovery/environment/native/src/format/thousands_group.cpp
#include "freight/formats.h"

#include <algorithm>
#include <cstdio>

namespace freight {

// Formatter: thousands group.
std::string fmt_thousands_group(long long value) {
  bool negative = value < 0;
  long long absolute = negative ? -value : value;
  const std::string digits = std::to_string(absolute);
  std::string grouped;
  int count = 0;
  for (int i = static_cast<int>(digits.size()) - 1; i >= 0; --i) {
    grouped.push_back(digits[static_cast<size_t>(i)]);
    ++count;
    if (count % 3 == 0 && i > 0) {
      grouped.push_back(',');
    }
  }
  std::reverse(grouped.begin(), grouped.end());
  return negative ? "-" + grouped : grouped;
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/format/ThousandsGroup.java
package com.freight.format;

/** Formatter: thousands group. */
public final class ThousandsGroup implements Formatter {

    private static final String BASE36_DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    @Override
    public String name() {
        return "thousands_group";
    }

    @Override
    public String apply(long value) {
        boolean negative = value < 0L;
        long absolute = negative ? -value : value;
        String digits = Long.toString(absolute);
        StringBuilder grouped = new StringBuilder();
        int count = 0;
        for (int i = digits.length() - 1; i >= 0; i--) {
            grouped.append(digits.charAt(i));
            count++;
            if (count % 3 == 0 && i > 0) {
                grouped.append(',');
            }
        }
        grouped.reverse();
        return (negative ? "-" : "") + grouped;
    }
}
